File: baekjoon/예산.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int n;
int arr[10001];
int total;

int main(){
    cin>>n;
    int answer=0;
    for (int i = 0; i < n; ++i) {
        cin>>arr[i];
    }
    cin>>total;
    sort(arr, arr + n);

    int start=0;
    int end = arr[n-1];
    while (start <= end) {
        int sum=0;
        int mid = start + (end-start)/2;
        for (int i = 0; i < n; ++i) {
            sum += min(arr[i], mid);
        }
        if (sum > total) {
            end = mid-1;
        }else{
            answer = mid;
            start = mid +1;
        }
    }

    cout<<answer;
}

File: baekjoon/예산.py
import sys

if __name__ == '__main__':
    t = int(sys.stdin.readline())
    arr = list(map(int, input().split()))
    m = int(sys.stdin.readline())

    arr.sort()

    start = 1
    end = arr[t - 1]
    mid = 0
    answer = 0
    while start <= end:
        mid = (start + end) // 2
        answer = mid
        sum = 0
        for i in range(t):
            if mid > arr[i]:
                sum += arr[i]
            else:
                sum += mid
        if sum > m:
            end = mid - 1
            answer = end
        else:
            start = mid + 1
    print(answer)
